### ISSUE TITLE
Need new rsk data set

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#713]
With recent changes to all the `rsk-class` functionality, it is time to replace the sample `rsk` dataset with something more modern. Ideally this should be an actual `.rsk` file, so that the full suite of SQLite interactions can be used and tested at build. The issue currently is that most `.rsk` files are too large to be included as a standard dataset and must be compressed (and trimmed of "unnecessary" tables). This may soon be able to be solved by improvements on the Ruskin side, which will allow export of a specific region of an rsk to a new rsk. If such improvements happen it would be easy to get a smaller rsk file.

I would also propose that we could include a second example dataset, which is an ASCII text export of the same `.rsk` file. That way, tests can be incorporated into the package build to compare the reading of the two raw files to make sure that changes to the code for one or the other don't break things.

Thoughts @dankelley?

**Comment by dankelley** [comment]
I like this but with a caution on file size. We don't want to get into CRAN hassles, and we don't want to get into a position where we cannot add new data to oce because it's already at an amber-light size.

Quite a lot of trimming has been done on other datasets, and if we were to perhaps add several 100K, my idea would be to shove ``coastlineMedium`` in there, because in a lot of coastal work the default coastline is pretty coarse. (I saw this coarseness when some mar-bio students came by for help. They didn't know about ``ocedata`` and their NS coastlines were so blocky I didn't realize they were using local data)

 It's not clear to me that including data just for build tests is correct. The build tests are to ensure that code works across platforms and versions of R. But I don't think we are going to have such issues ... once we know it tests ok on one box, it will probably be ok on others. So maybe we should have separate tests. We can do that within the oce repo, in a directory (like ``create_data``) that doesn't go into the oce package.

**Comment by dankelley** [comment]
I'd be inclined to mark this with 4 stars (meaning no new release until fixed) ... @richardsc ok?

**Comment by richardsc** [comment]
I've got lots of good candidate files for this now, but as predicted the problem will be getting something small enough that it will be suitable for inclusion. For a typical profiling instrument (sampling at say, 4Hz), and a profile that takes 30 mins or so, the file size is at least 0.5 MB. There are some tables that can be removed, but it's not enough to bring it down to ~50 KB.

One option would be to trim the rsk data table directly using sqlite, but I really don't know what I'm doing. I tried doing it manually with a sqlite editor, but discovered that the data are not written to the table chronologically, so trimming large blocks ends up producing a bit of a mess.

Any thoughts? I like the thought of producing a sample file from a "true" rsk using a script that can be re-run, but I don't have the skills for it.

**Comment by dankelley** [comment]
A question is whether we are talking about including a ``.rsk`` file (in ``inst/extdata``) or an object created from one. If the latter, and if oce is up to it, we could use ``subset`` to get a smaller ``.rda`` file. If the former, I think we need to drop the idea unless RBR has well-tested software to subset from a dataset.  I fear that if we manually tweak the ``.rsk`` file, we may get something wrong, and that won't serve much purpose (put another way, it could mess things up).

For half a day, a week ago, I had a ``data/odf.rda`` object in there. Then I realized there was really not much point since it's just a data format, not really a data type.  I wonder if the same may be true of Ruskin format.  Put another way, if the only use to which these datasets would be put would be e.g.
```r
data(rsk)
plot(rsk)
new <- subset(rsk, pressure<100)
```
or similar, then I don't see any huge advantage to having the data built in like that. Users who do not have ``.rsk`` data will gain nothing. And those with ``.rsk`` data will be doing the nearly-equivalent
```r
rsk <- read.oce("myDataFile.rsk")
plot(rsk)
new <- subset(rsk, pressure<100)
```
Also, if RBR has a stable URL on a sample file, then that could be used directly too.  (**Question:** does RBR have such a URL?)

In other words, I'm coming around to saying we simply give up on including an ``rsk`` dataset. We already have a class (``odf``) that has no built-in dataset.  Maybe we should restrict these built-in datasets to things that are the most useful for learning (I would argue ``ctd`` falls into that category, and also ``sealevel`` and ``section``) or for work (here we have ``coastlineWorld`` and similar).

PS. as regards your last line, I think you mean including an ``rsk`` in ``inst/extdata`` but I'd suggest we do not do that. I am thinking of removing the ``odf`` file that's there, for example, because the only people who would ever want to use ``odf`` functions are people who have odf datasets.

**Comment by dankelley** [comment]
Another thought -- I don't think GH has stringent restrictions on file size, and a 1Mb file could go into the oce webpage. We could have a subpage called http://dankelley.github.io/oce/datasets.html for example. 

That webpage could be a place to put the all the sample files in ``inst/extdata``, saving us 244Kb in the (uncompressed) oce size and I think about 62Kb in the compressed tarball size.

I don't see any negative to this, actually, if GH has no policy against it. I don't think it would be right to start putting tons of stuff there, e.g. the whole levitus atlas or whatever, because it's better for users to download their own (and thus up-to-date) data.

**Comment by dankelley** [comment]
Another thought ... on that new GH webpage we could have a link table, e.g. if you want levitus data, click it's link. If you want landsat, click another link. I would use that, myself.

**Comment by richardsc** [comment]
I think those are both great ideas. Would be nice if we could link to such pages in the man pages, so that people reading about classes can go find the examples easily.

**Comment by dankelley** [comment]
Another another thought (sorry) -- if that table were written right, oce could scan it. So there could be a function like ``lookupDataset("woa")`` which could return a URL for Levitus.

**Comment by richardsc** [comment]
Just catching up with this now.

I agree that it probably isn't worth trying to hack a file small enough that it could be used in `inst/extdata` since that will most likely lead to issues down the road. I like the idea of including some "extra" example dataset on the github webpage, which we can use for testing or pointing new users to.

I do, however, like the idea of having an `rsk` object included in oce (as it is now), even if the raw data (that created it) isn't present in the package. That way we can keep the storage to a minimum, but still use it to test and highlight differences between `rsk` class and other related classes (e.g. `ctd`).

So I think this Issue is still relevant as it stands, but the end goal is not to include an RSK file but rather an `rsk` object created from one.

**Comment by richardsc** [comment]
I've started work on this and will have a new rda file for the next release (before the end of this year).

**Comment by richardsc** [comment]
Any particular limit for file size? After a rough trim I get an rda that is 64K. I could further trim it by decimating, but I'd rather leave the data in the raw sample rate if possible (because otherwise the metadata is wrong).

**Comment by dankelley** [comment]
I think 64K is in the right ballpark. Is it a ctd-style, and are you trimming to the downcast only? If so, then that seems like a natural file configuration. NB some of the other files can be trimmed if need be. For example, landsat is 98K with `decimate(..., by=100)` and I guess it could be made to be 30K or so with `by=150`, which means the rsk one would add only 30K to the total. Also we don't really need sealevelTuktoyaktuk.rda (7K) or ctdRaw.rda (20K), so we could get the whole package to be about the same size as before, I think.  (I just don't want to get into arguments with you-know-who, and if oce gets a bit bigger but we've seriously trimmed it to moderate the increase, I think we will be okay.)

**Comment by richardsc** [comment]
> @dankelley wrote: I think 64K is in the right ballpark. Is it a ctd-style, and are you trimming to the downcast only? If so, then that seems like a natural file configuration.

It is a ctd-style instrument, and I'm currently trimming to downcast only. I could take a bit off the bottom to make it a bit smaller. I'd rather not remove other datasets though -- `sealevelTuktoyaktuk` is a good example for `tidem()` because it's the classic example from both Foreman and Pawlowicz, and `ctdRaw` is pretty useful for demonstrating ctd trimming, etc. Maybe we could trim `landsat` a bit ...

**Comment by dankelley** [comment]
We used to have a landsat that was a localized region (Cape Split) and I may go back and see how big that was.  I think a full-resolution of a small region might be quite interesting.

I agree on keeping `sealevelTuktoyaktuk`. Actually, I think I'm worried about nothing. None of these files is particularly large, and we don't have tons of them, either. I suspect we are being as reasonable as anybody else.

**Comment by richardsc** [comment]
Sorry to have let this go so stale, but I think I'm close to being done. The one question I have is:

Will we include an *actual* RSK file, or just an `rsk` object (in `.rda` form) created from an RSK file?

For the data I chose, the original RSK file is 834K, but the trimmed and re-saved rda file (after using `tools::resaveRdaFiles()`) is only 544B.

**Comment by dankelley** [comment]
You mean 544K, right?

I think that 0.8Mb rsk file is too big to add into the package ... it would increase the size by 50% or whatever.  And I don't think users will need it in their own work.  I somehow doubt that anyone uses the things in `inst/extdata`) at the moment, and think we should remove them from the CRAN package (for elegance, not size, considerations, since they will compress in the tarball).

At one time, the build tests would have seemed to benefit from things in `inst/extdata`, but I cannot recall a case where CRAN tests found an error in the data that local tests would not find.  And we can set up our new `testthat` tests to have components that are skipped on CRAN, so if we keep datasets in `inst/extdata` but set up `.Rbuildignore` to skip them, it means we get the tests we want but without making the package large. That seems like the solution to any complaints from CRAN.  And I'm pretty sure they would complain about increasing the size by 834K for a file that is not really needed.  (If we wanted to increase by that amount, we could add in `coastlineMedium` or some other thing that is likely to help a high percentage of users.)

**Comment by richardsc** [comment]
No, I mean that the *final* rda file is only 544 Bytes:
```
$ ll
-rw-------@ 1 richardsc  staff   834K 27 Nov 21:03 060130_20150904_1159.rsk
-rw-r--r--@ 1 richardsc  staff   302B 14 Dec 19:22 makeRskData.R
-rw-r--r--@ 1 richardsc  staff   1.0K 14 Dec 19:22 makeRskData.out
-rw-r--r--@ 1 richardsc  staff   544B 14 Dec 19:22 rsk.rda
```
and the script that makes the final rda is:
```r
rm(list=ls())
library(oce)
raw <- read.oce('060130_20150904_1159.rsk')
focus <- structure(c(1441381020.7302, 1441381482.90486), class = c("POSIXct", 
"POSIXt"), tzone = "UTC")
raw <- subset(raw, focus[1] <= time & time <= focus[2])
rsk <- raw
save(file='rsk.rda', rsk)
tools::resaveRdaFiles("rsk.rda")
```

**Comment by dankelley** [comment]
Whoa, that's tiny! Amazing that this can contain much data.  Or maybe that `resaveRdaFiles` is amazing.

Q: what do you think of my argument that we not include the `*.rsk` file?

**Comment by richardsc** [comment]
I'm fine with that. It means the original RSK will still be required to regenerate the `rsk` object, but I guess you've got a setup now that organizes such things?

**Comment by dankelley** [comment]
Yup, you do
```
cd oce/create_data/rsk
make clean
make
```
and after that it makes a lot of sense to open R in the dir and `load` the file and do a lot of tests. As I regenerated all the files for units and flags, I compared against the existing files, and in quite a bit of detail.  Once a person is *really* sure it's OK, then do
```
make install
```
which copies it to ``../../data``.  Then do a "build check" in RStudio or otherwise, and when that seems OK, then probably it is ok to upload to GH.

We now have a subdir in `create_data` for every data type.  It took me quite a long time to get things set up but I'm glad I did because a lot of things were done manually before and it's terribly easy to forget where the data came from when you do that.

NOTE: `create_data` is in `.Rbuildignore`, so it does not get put into the R tarball.

**Comment by richardsc** [comment]
Ok, I think this is done. The final rda file is actually 48k. I will email the raw file and the generation script to @dankelley so that it can be incorporated in the proper way (not trying to get out of doing work here, just I think it would be faster for you to do it).

It will also need documentation. Is there an Roxygen way of documenting data sets, or should I just edit the current `rsk.Rd` file?

**Comment by dankelley** [comment]
The new rsk is in `develop` now. Also, there is a template for the roxygen2 doc for it, in `R/rsk.R`. I removed the old `man/rsk.Rd` file. The easiest way to make the new doc is within Rstudio; just hit command-shift-D in the project.  You can also run roxygen by hand, but I did that the other day and it clobbered `DESCRIPTION`.  So the best is to use Rstudio.

A neat feature that I only learned of lately is `@family`, e.g. I'm putting satellite stuff in a family. This basically creates 'see also' entries automatically, which is saves work for the programmer and makes things an awful lot more helpful to users.  I will be making a new family for adv and adp.  I don't know if family requires roxygen, though ... I think it does. But adv is already in roxygen, and I think adp is about halfway there, so I'll complete that on the weekend.

**Comment by dankelley** [comment]
@richardsc if you like, I can add two paragraphs to the docs in `R/rsk.R`.  Ideally, it would mention whether that top point is 'real', since it looks wonky on the TS diagram (to me, anyhow).  Or maybe that point could be chopped?

<img width="568" alt="screen shot 2016-01-30 at 1 55 12 pm" src="https://cloud.githubusercontent.com/assets/99469/12697424/371ec73a-c759-11e5-9041-218813efdd37.png">

**Comment by richardsc** [comment]
Sorry for being slow -- I plan to get to the documentation ASAP. And I agree, that top point should probably be trimmed. I'm not sure why I didn't already.

**Comment by richardsc** [comment]
Ok, I added some documentation. I pushed it straight to develop. Note that commit 96ef8bff0706ba8a77 changed quite a few things after I ran roxygen in RStudio.

**Comment by dankelley** [comment]
Thanks! I just pushed an updated, with all those other files changed back. You have an old version of roxygen, which inserts its version number in each .Rd file. This was a total pain because every time roxygen was developed, all the `man/*Rd` files changed ... for people using version control, this is not desirable.  But I have the new roxygen so when I used it, all those files went back to normal and my diff just showed the two paragraphs. I'm going to close this, to save comments back and forth on whether you agree. As stated in the title, it's been addressed.  However, I'm opening up a new issue #848 regarding the wild point near the top.
